### PR TITLE
PERF-#6437: preserve dtypes for 'reindex'

### LINKED ISSUE #6437: Preserve dtypes cache for `reindex`

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6437 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:653]: What is the magic number equal to -1? Let's assing it to a descriptive var.

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:655]: Move to the top if this method is really needed.

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:650]: Index is a simpler entity than Series, and it suffices here.

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:658]: I'm not sure, but a few parts of the code still use the function from pandas (so I decided to use it too).

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (vnlitvinov) [inline modin/core/storage_formats/pandas/query_compiler.py:657]:
```suggestion
                    for col, col_dtype in new_dtypes.items():
                        new_dtypes[col] = find_common_type((col_dtype, dtype))
```

**Review by vnlitvinov — COMMENTED** [review]
Overall looks good, not stamping approval because I still would like to see the in-code comment about columns vs rows added first 😺

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:644]: > Please add this as an in-code comment 🙃

@vnlitvinov added

**Review by vnlitvinov — APPROVED** [review]
LGTM! Let's wait for @YarShev feedback before merging, though, because he also left some questions and other feedback.

**Review by YarShev — APPROVED** [review]

**Pull request merged**
